### PR TITLE
[multikueue] Decrease the number logged errors

### LINKED ISSUE #1814: [MultiKueue] in around 30min intervals error message is logged during normal operation
**What happened**:

During normal operation MultiKueue logs errors repeatedly every 30min per cluster:
```
{"level":"error","ts":"2024-03-06T08:21:44.586427722Z","caller":"multikueue/multikueuecluster.go:200","msg":"Cannot get workload key","clusterName":"multikueue-test-worker2","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet","jobKind":"/, Kind=","error":"not a jobset","stacktrace":"sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.(*remoteClient).startWatcher.func1\n\t/workspace/pkg/controller/admissionchecks/multikueue/multikueuecluster.go:200"}
```
Longer example here: https://github.com/kubernetes-sigs/kueue/pull/1806#issuecomment-1980342324.

**What you expected to happen**:

No error logs reported during normal operation.

**How to reproduce it (as minimally and precisely as possible)**:

Run MultiKueue as described in https://github.com/kubernetes-sigs/kueue/issues/1787. Tested on GKE

**Anything else we need to know?**:

This is a follow up to https://github.com/kubernetes-sigs/kueue/issues/1787

**Environment**:
- Kubernetes version (use `kubectl version`): 1.28.3-gke.1286000
- Kueue version (use `git describe --tags --dirty --always`): master build
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Limit the number and severity of logged errors during normal operations.
#### Which issue(s) this PR fixes:

Fixes #1814 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 1]
```diff
@@ -180,7 +180,10 @@ func (w *wlReconciler) remoteClientsForAC(ctx context.Context, acName string) (m
 	clients := make(map[string]*remoteClient, len(cfg.Spec.Clusters))
 	for _, clusterName := range cfg.Spec.Clusters {
 		if client, found := w.clusters.controllerFor(clusterName); found {
-			clients[clusterName] = client
+			// Skip the client if its reconnect is ongoing.
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:183]: Some questions:
1. Is this related to the errors, or drive by fix? 
2. What is the user-visible impact (if any)?

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 2]
```diff
@@ -196,11 +197,21 @@ func (rc *remoteClient) startWatcher(ctx context.Context, kind string, w multiKu
...
-				log.V(2).Error(err, "Cannot get workload key", "jobKind", r.Object.GetObjectKind().GroupVersionKind())
-			} else {
-				rc.queueWorkloadEvent(ctx, wlKey)
+			switch r.Type {
+			case watch.Error:
+				switch s := r.Object.(type) {
+				case *metav1.Status:
+					log.V(5).Info("Watch error", "status", s.Status, "message", s.Message, "reason", s.Reason)
+				default:
+					log.V(5).Info("Watch error with unexpected type", "type", fmt.Sprintf("%T", s))
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:206]:
```suggestion
					log.V(2).Error("Watch error with unexpected type", "type", fmt.Sprintf("%T", s))
```
If this is unexpected, we prefer to get it logged at high level.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 1 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/workload.go:183]: I't not the error in the issue, but if the watch has failed and the client tries to reconnect and the, say network connection is still down, trying to get workloads from the worker will also fail. So in my opinion we should skip  that client for the time being.

Alternatively we can just skip the reporting of the error in `readGroup`, but it looks strange.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 2 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:206]: The type of the error is unexpected, the fact that the error is reported is as expected as two lines above.

Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:206]: The message "Watch error with unexpected type" suggests something is going seriously wrong, so setting (5) things going badly. WDYT?

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:183]: IIUC there is no user-facing impact except for skipping some requests which are likely to fail. 

Could we have this maybe as a separate PR if not related, and with some unit test?

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 2 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:206]: Oh, I see you don't have err at hand here to call `Error`, still, v(2) seems more appropriate

Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go]: Let's go with V(3)

Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go]: Let's go with V(3)

**Comment by mimowo** [comment]
/lgtm
/approve
Thanks

**Comment by alculquicondor** [comment]
Can you post what the "errors" looked like?

Are they completely ignorable?

**Comment by mimowo** [comment]
> Can you post what the "errors" looked like?

The main one we care about is under the issue. It is returned when the watch expires:
```
{"level":"error","ts":"2024-03-06T08:21:44.586427722Z","caller":"multikueue/multikueuecluster.go:200","msg":"Cannot get workload key","clusterName":"multikueue-test-worker2","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet","jobKind":"/, Kind=","error":"not a jobset","stacktrace":"sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.(*remoteClient).startWatcher.func1\n\t/workspace/pkg/controller/admissionchecks/multikueue/multikueuecluster.go:200"}
```

**Comment by mimowo** [comment]
The other one "log.V(3).Info("Watch error with unexpected type", "type", fmt.Sprintf("%T", s))" I never observed

**Comment by mimowo** [comment]
^^ This is before the change. I don't have ready-to-paste output after the change. I could generate one easily.

**Comment by mimowo** [comment]
This is how kueue logs now:

```
> kubectl logs kueue-controller-manager-5b48c8b59b-wg57n -nkueue-system 2>&1 | grep -i watch
{"level":"info","ts":"2024-03-13T14:38:41.164876413Z","logger":"controller-runtime.certwatcher","caller":"certwatcher/certwatcher.go:161","msg":"Updated current TLS certificate"}
{"level":"info","ts":"2024-03-13T14:38:41.165088008Z","logger":"controller-runtime.certwatcher","caller":"certwatcher/certwatcher.go:115","msg":"Starting certificate watcher"}
{"level":"Level(-2)","ts":"2024-03-13T14:39:05.267615703Z","caller":"multikueue/multikueuecluster.go:198","msg":"Starting watch","clusterName":"multikueue-test-worker2","watchKind":"Workload.kueue.x-k8s.io"}
{"level":"Level(-2)","ts":"2024-03-13T14:39:05.27162791Z","caller":"multikueue/multikueuecluster.go:198","msg":"Starting watch","clusterName":"multikueue-test-worker2","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet"}
{"level":"Level(-2)","ts":"2024-03-13T14:39:05.285750407Z","caller":"multikueue/multikueuecluster.go:198","msg":"Starting watch","clusterName":"multikueue-test-worker1","watchKind":"Workload.kueue.x-k8s.io"}
{"level":"Level(-2)","ts":"2024-03-13T14:39:05.290619177Z","caller":"multikueue/multikueuecluster.go:198","msg":"Starting watch","clusterName":"multikueue-test-worker1","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet"}
{"level":"Level(-2)","ts":"2024-03-13T15:12:27.159711619Z","caller":"multikueue/multikueuecluster.go:217","msg":"Watch ended","clusterName":"multikueue-test-worker1","watchKind":"Workload.kueue.x-k8s.io","ctxErr":null}
{"level":"Level(-2)","ts":"2024-03-13T15:12:27.159799974Z","caller":"multikueue/multikueuecluster.go:223","msg":"Queue reconcile for reconnect","clusterName":"multikueue-test-worker1","watchKind":"Workload.kueue.x-k8s.io","cluster":"multikueue-test-worker1"}
{"level":"Level(-3)","ts":"2024-03-13T15:12:27.160910077Z","caller":"multikueue/multikueuecluster.go:204","msg":"Watch error","clusterName":"multikueue-test-worker1","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet","status":"Failure","message":"an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding","reason":"InternalError"}
{"level":"Level(-2)","ts":"2024-03-13T15:12:27.161043184Z","caller":"multikueue/multikueuecluster.go:217","msg":"Watch ended","clusterName":"multikueue-test-worker1","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet","ctxErr":"context canceled"}
{"level":"Level(-2)","ts":"2024-03-13T15:12:27.167277052Z","caller":"multikueue/multikueuecluster.go:198","msg":"Starting watch","clusterName":"multikueue-test-worker1","watchKind":"Workload.kueue.x-k8s.io"}
{"level":"Level(-2)","ts":"2024-03-13T15:12:27.171478071Z","caller":"multikueue/multikueuecluster.go:198","msg":"Starting watch","clusterName":"multikueue-test-worker1","watchKind":"jobset.x-k8s.io/v1alpha2, Kind=JobSet"
```
I guess not too bad, note that now the message "Watch error" is `V(3)`. WDYT?

**Comment by alculquicondor** [comment]
I don't like the idea of hiding errors as V(3) Info. If they are expected, I have a hard time understanding why they are marked as `InternalError`.

Maybe @apelisse can provide some guidance?

**Comment by alculquicondor** [comment]
What context are we passing to start the watch? I wonder if controller-runtime has a timeout for the context.

**Comment by mimowo** [comment]
@trasc can you help to answer that?

**Comment by mimowo** [comment]
I looked at this a little bit more, here is the summary:
* I confirmed that the watch is closed depending on the `min-request-timeout` param. When set to 30s the watch usually ends in 30s - 1min. This is consistent for kind and GKE.
* The watch ends with 200 status code, there is no error server side. Here is an example server-side log by the API server, when the watch was closed after 34s:
```
1 httplog.go:132] "HTTP" verb="WATCH" URI="/apis/jobset.x-k8s.io/v1alpha2/jobsets?labelSelector=kueue.x-k8s.io%2Fmultikueue-origin%3Dmultikueue&watch=true" latency="34.525929114s" userAgent="manager/v0.0.0 (linux/amd64) kubernetes/$Format" audit-ID="55d611cc-e3b5-4643-b5b1-92182703a262" srcIP="192.168.8.1:36440" apf_pl="workload-low" apf_fs="service-accounts" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_init_latency="320.89µs" apf_execution_time="323.26µs" resp=200
```

I think I traced down a little more what happens at the client side.
1. the client creates a new watcher with InternalError (500) as the default: https://github.com/kubernetes/kubernetes/blob/03ce04584437624840ad78edac1b772e47e78dc2/staging/src/k8s.io/client-go/rest/request.go#L789-L791
2. the watcher wraps the closed connection with the error that we see in the logs: https://github.com/kubernetes/kubernetes/blob/03ce04584437624840ad78edac1b772e47e78dc2/staging/src/k8s.io/apimachinery/pkg/watch/streamwatcher.go#L120
3. this is further wrapped by the `AsObject` function which delegates to `NewGenericServerResponse`: https://github.com/kubernetes/kubernetes/blob/03ce04584437624840ad78edac1b772e47e78dc2/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L843-L857
4. The `NewGenericServerResponse` does the final wrapping (because 500 is the defualt): https://github.com/kubernetes/kubernetes/blob/03ce04584437624840ad78edac1b772e47e78dc2/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L487-L490

This seems also reported here: https://github.com/kubernetes/client-go/issues/12, and the suggestion seems to just break the loop when the channel is closed, regardless of the event. 

Finally. handling of the nuances of watcher recreation seems to be already covered by the `RetryWatcher` here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/watch/retrywatcher.go#L149-L236. WDYT @alculquicondor? I can open a follow up issue to investigate / clean up.

**Comment by mimowo** [comment]
Sg, closing the cherry pick. I will open the issue for client-go. I'm yet wondering about using the RetryWatcher to abstract out this craft. Any view, I think it makes sense to open an issue, maybe someone will pick this up.

**Pull request merged**